### PR TITLE
[OPS-963] Switch to opam-nix

### LINKED ISSUE #44: [OPS-963] Use opam-nix
# Clarification and motivation
There is a tool https://github.com/balsoft/opam-nix, which worked nice for compiling LIGO, we should
try it here as well.

# Acceptance criteria
`opam-nix` is used for tezos compiling.

### PR DESCRIPTION
**Pull request by balsoft**
## Description

Manually packaging all of the dependencies is hard and requires a lot of human effort. To solve this, I'm using [opam-nix](https://github.com/balsoft/opam-nix). 

## Related issue(s)

Resolves #44

#### Related changes (conditional)

- [x] I checked whether I should update the [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).

### CODE REVIEW

File: .#default.nix [hunk 1]
```diff
@@ -0,0 +1 @@
+balsoft@AMD-Workstation.24149:1584705302
```
Comment (rvem) [inline .#default.nix:1]: :troll:

Comment (balsoft) [reply .#default.nix]: I USE EMACS BTW

Comment (zhenyavinogradov) [reply .#default.nix]: :coolstory:

**Comment by balsoft** [comment]
So, noone is willing to actually review a 10-page diff?

I'm not surprised.

File: scripts/autorelease.sh [hunk 2]
```diff
@@ -43,4 +41,4 @@ for file in $assets_dir/*; do
 done
 
 # Create release
-hub release create "${assets[@]}" -F "$TEMPDIR"/"$project"/*-release-notes.md --prerelease auto-release
+hub release create "${assets[@]}" -F "$TEMPDIR"/"$project"/release-notes.md --prerelease auto-release
```
Comment (rvem) [inline scripts/autorelease.sh:44]: Have you tested releasing, btw?

File: release-binaries.nix [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# SPDX-FileCopyrightText: 2019 TQ Tezos <https://tqtezos.com/>
+#
+# SPDX-License-Identifier: MPL-2.0
+
+let
+  protocols = import ./protocols.nix;
+  protocolsFormatted = builtins.concatStringsSep ", " (protocols.allowed ++ protocols.active);
+in[
+  {
+    name = "tezos-client";
```
Comment (rvem) [inline release-binaries.nix:10]: now it also has `tezos-admin-client` binary, which is weird

File: debSource.nix [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    runCommand "build_source_deb" { SOURCE_DATE_EPOCH = 1576846270; } ''
+      mkdir -p ${project}/debian/source
+      cd ${project}
+      cp ${writeMakefile} Makefile
+      cp ${pkg}/bin/* .
+      cp ${writeControlFile} debian/control
+      cp ${writeChangelogFile} debian/changelog
+      cp ${meta.licenseFile} debian/copyright
+      echo 9 > debian/compat
+      echo "The Debian Package tezos-client-mainnet" > debian/README
```
Comment (rvem) [inline debSource.nix:80]: this should contain the specific binary name, probably I forgot to do it a long time ago :(

File: README.md [hunk 5]
```diff
@@ -18,27 +18,10 @@ Recomended way to get these binaries is to download them from assets from github
...
-Contents of release:
-* `tezos-*-005-PsBabyM1` static binaries for 005 protocol.
-* `tezos-*-006-PsCARTHA` static binaries for 006 protocol.
-* `tezos-*` static binaries that can be used with any protocol.
-* `binaries-<revision>.tar.gz` archive with all binaries made from
-particular master branch revision.
-* License file from [tezos repository](https://gitlab.com/tezos/tezos/).
+Some of the individual binaries contain protocol name to determine
+with which protocol binary is compatible with. If this is not the
+case, then consult release notes to check which protocols are
```
[the remark this answers is not in the document]
Comment (balsoft) [reply README.md:23]: I'm disabling all the protocols that are in `ignored`.

File: scripts/autorelease.sh (hunk 2 above)
Comment (balsoft) [reply scripts/autorelease.sh:44]: No.

File: release-binaries.nix (hunk 3 above)
Comment (balsoft) [reply release-binaries.nix:10]: Well, it's built as part of `tezos-client` opam package

File: debSource.nix (hunk 4 above)
Comment (balsoft) [reply debSource.nix:80]: Oof

File: protocols.json [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    "genesis-carthagenet",
+    "demo-counter",
+    "000-Ps9mPmXa",
+    "001-PtCJ7pwo",
+    "002-PsYLVpVv",
+    "003-PsddFKi3",
+    "004-Pt24m4xi"
+  ],
+  "allowed": [
+    "005-PsBABY5H",
```
[the remark this answers is not in the document]
Comment (balsoft) [reply protocols.json:14]: You can `./proto ignore 005-PsBABY5H` and push here yourself :)

**Comment by balsoft** [comment]
@rvem I should've made it clearer.
- `ignored` protocols are not built at all, and binaries like `tezos-client` don't include support for them.
- `allowed` protocols are built into binaries like `tezos-client`, but protocol-specific binaries (like `tezos-accuser`) are not built for them. Realistically, it's only used in changelogs and descriptions, because this is kind of the default.
- `active` protocols are fully active -- protocol-specific binaries are built, and support for those protocols is included everywhere it should be included.

File: scripts/autorelease.sh (hunk 2 above)
Comment (balsoft) [reply scripts/autorelease.sh:44]: Tested a release, see here: https://github.com/serokell/tezos-packaging/releases/tag/auto-release

Comment (rvem) [reply scripts/autorelease.sh:44]: image
looks weird, I won't say that `006-PsCARTHA` binaries target the 005 protocol

Comment (balsoft) [reply scripts/autorelease.sh:44]: Well, this is just wording. Binaries that don't have a suffix do indeed target 005 and 006. This is expressed in the updated README.

Comment (balsoft) [reply scripts/autorelease.sh:44]: Updated the changelog.

**Comment by balsoft** [comment]
@rvem please re-evaluate and supply more feedback (if any).

File: debSource.nix (hunk 4 above)
Comment (balsoft) [reply debSource.nix]: Flxed

**Comment by balsoft** [comment]
@rvem could you check if I packaged `dput` correctly and it works for you? If so, we could add a script that would automatically upload stuff to launchpad.

File: pkgs.nix [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# SPDX-FileCopyrightText: 2019 TQ Tezos <https://tqtezos.com/>
+#
+# SPDX-License-Identifier: MPL-2.0
+
+{ sources ? import ./nix/sources.nix, pkgs ? import sources.nixpkgs { } }:
+let
+  ocaml-overlay = import ./ocaml-overlay.nix { inherit sources pkgs; };
+  static-overlay = import ./static-overlay.nix;
+  normal = import sources.nixpkgs { overlays = [ ocaml-overlay ]; };
```
Comment (zhenyavinogradov) [inline pkgs.nix:9]: Why do you extend nixpkgs with `ocaml-overlay` here? It looks like it does not affect anything

Comment (balsoft) [reply pkgs.nix:9]: This is for easier testing of your changes, I guess. You can do `nix build -f pkgs.nix ocamlPackages.tezos-client` and you will have to build a lot less stuff.

TBH, I initially thought we need a dynamic version of all the executables, but we don't. So maybe this is not needed indeed.

Comment (zhenyavinogradov) [reply pkgs.nix:9]: Ok, keeping possibility to build dynamic executables is good, even if we don't actually use it right now

Comment (balsoft) [reply pkgs.nix]: Restructured `pkgs.nix` to make it extra clear what's going on.

File: .buildkite/pipeline.yml [hunk 8]
```diff
@@ -15,12 +15,12 @@ steps:
    label: crossref-verify
    soft_fail: true
  - commands:
-   - nix-build -A binaries -o binaries
-   - nix-build -A deb-packages -o deb-packages --arg timestamp ${cur_date}
-   - nix-build -A rpm-packages -o rpm-packages --arg timestamp ${cur_date}
+   - "nix-shell -p gnutar --run 'nix-build -A binaries -o binaries'"
```
Comment (zhenyavinogradov) [inline .buildkite/pipeline.yml:18]: `nix-shell` here fixes the problem with `bzip2` missing, but I think it's better to add `bzip2` to our agent environment, as nix relies on it in run-time

**Comment by zhenyavinogradov** [comment]
Other than that, everything looks good

File: .buildkite/pipeline.yml (hunk 8 above)
Comment (balsoft) [reply .buildkite/pipeline.yml:18]: This is obviously just a workaround, so yes, ideally we should fix our bk agent.

Comment (zhenyavinogradov) [reply .buildkite/pipeline.yml:18]: Will you make a PR for our agent config?

Comment (balsoft) [reply .buildkite/pipeline.yml:18]: Interestingly, `gnutar` is already part of `runtimePackages`  for both of our agents.

Comment (balsoft) [reply .buildkite/pipeline.yml:18]: Ah, I think I see the problem, right

Comment (balsoft) [reply .buildkite/pipeline.yml:18]: https://github.com/serokell/serokell-profiles/pull/457

Comment (balsoft) [reply .buildkite/pipeline.yml:18]: Added a PR.

**Comment by balsoft** [comment]
Let's merge this thing! Please re-evaluate it. I'd like this to be the basis for implementing #47

File: release-binaries.nix (hunk 3 above)
Comment (rvem) [reply release-binaries.nix:11]: I see, but still, I'm not sure if it's a good idea to break existing package per binary principle

Comment (balsoft) [reply release-binaries.nix:11]: Do I just `rm tezos-admin-client` in `postInstall` or do I create a separate package for it?

Comment (rvem) [reply release-binaries.nix:11]: Let's have a separate package for it as it was done previously

Comment (balsoft) [reply release-binaries.nix:11]: @rvem done

**Comment by balsoft** [comment]
@rvem please check that this is what you want. Also, I'd appreciate it if you tested actually installing debs and rpms (I haven't)

**Comment by rvem** [comment]
I've tested both `*.deb` and `*.rpm` packages for `tezos-node`, they seem to be fine

**Comment by balsoft** [comment]
@rvem cool! I'll autosquash and then merge if you don't mind.

**Comment by rvem** [comment]
sure

**Pull request merged**